mlx-fast/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+    let base = mlx_sys::mlx_optional_float {
+        value: base.unwrap_or(0.0),
+        has_value: base.is_some(),
+    };
+
+    unsafe {
+        let new_array = mlx_sys::mlx_fast_rope(
+            array.as_ref().as_ptr(),
+            dimensions,
+            traditional,
+            base,
+            scale,
+            offset,
+            freqs
+                .into()
+                .map(|a| a.as_ptr())
+                .unwrap_or(std::ptr::null_mut()),
+            stream.as_ref().as_ptr(),
+        );
+
+        Array::from_ptr(new_array)
+    }
+}
+
+/// A fast implementation of multi-head attention: `O = softmax(Q @ K.T, dim=-1) @ V`
+///
+/// Supports [Multi-Head Attention](https://arxiv.org/abs/1706.03762), [Grouped Query Attention](https://arxiv.org/abs/2305.13245), and [Multi-Query Attention](https://arxiv.org/abs/1911.02150).
+///
+/// This function will dispatch to an optimized Metal kernel when the query sequence length is 1. It handles other cases with regular MLX operations.
+///
+/// > Note: The softmax operation is performed in float32 precision regardless of input precision (float16 or float32).
+///
+/// > Note: For Grouped Query Attention and Multi-Query Attention, the input arrays for `key` and `value` should not be pre-tiled to match the `query` array.
+#[default_device]
+pub fn scaled_dot_product_attention_device<'a>(
+    queries: impl AsRef<Array>,
+    keys: impl AsRef<Array>,
+    values: impl AsRef<Array>,
+    scale: f32,
+    mask: impl Into<Option<&'a Array>>,
+    memory_efficient_threshold: impl Into<Option<i32>>,
+    stream: impl AsRef<Stream>,
+) {
+    let memory_efficient_threshold = memory_efficient_threshold.into();
+    let memory_efficient_threshold = mlx_sys::mlx_optional_int {
+        value: memory_efficient_threshold.unwrap_or(0),
+        has_value: memory_efficient_threshold.is_some(),
+    };
+
+    unsafe {
+        mlx_sys::mlx_fast_scaled_dot_product_attention(
+            queries.as_ref().as_ptr(),
+            keys.as_ref().as_ptr(),
+            values.as_ref().as_ptr(),
+            scale,
+            mask.into()
+                .map(|a| a.as_ptr())
+                .unwrap_or(std::ptr::null_mut()),
+            memory_efficient_threshold,
+            stream.as_ref().as_ptr(),
+        );
+    }
+}
+
+/// Root Mean Square normalization (RMS norm).
+///
+/// The normalization is with respect to the last axis of the input `x`.
+///
+/// # Params
+/// - x: input array
+/// - weight: A multiplicative weight to scale the result by. The `weight` should be one-dimensional with the same size as the last axis of `x`.
+/// - eps: A small additive constant for numerical stability
+/// - stream: stream or device to evaluate on
+#[default_device]
+pub fn rms_norm_device<'a>(
+    x: impl AsRef<Array>,
+    weight: impl AsRef<Array>,
+    eps: f32,
+    stream: impl AsRef<Stream>,
+) -> Array {
+    unsafe {
+        let new_array = mlx_sys::mlx_fast_rms_norm(
```

Comment:
We should catch error here right? The c++ side would throw if the shapes don't match wouldn't it?

Reply:
The same applies to other functions